<a href="https://colab.research.google.com/github/garcia-christian/DataAna_Activities/blob/main/MidtermExam_Garcia_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Part I

We need to import numpy and pandas before we can start our preprocessing

In [ ]:
import numpy as np
import pandas as pd

A.  using pd.read_excel we imported the ('Online Retail.xlsx') to retail_frame object

In [7]:
retail_frame = pd.read_excel('Online Retail.xlsx')

B. To identify missing values we used isnull() and used sum() to add the number of missing values.

In [10]:
print(retail_frame.isnull().sum())

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64


C we separated customerID and Description to other objects then we used fillna() to replace all the missing values with the mean and mode using mean() and mode() for the numerial and categorical data

In [ ]:
cus = retail_frame['CustomerID']
cus.fillna(cus.mean(),inplace = True)
desc = retail_frame['Description']
desc.fillna(desc.mode(),inplace=True)
retail_frame['CustomerID'] = cus
retail_frame['Description'] = desc

D.we used describe() to view the statistical details of retail frame

In [14]:
print(retail_frame.describe())

            Quantity      UnitPrice    CustomerID
count  541909.000000  541909.000000  541909.00000
mean        9.552250       4.611114   15287.69057
std       218.081158      96.759853    1484.74601
min    -80995.000000  -11062.060000   12346.00000
25%         1.000000       1.250000   14367.00000
50%         3.000000       2.080000   15287.69057
75%        10.000000       4.130000   16255.00000
max     80995.000000   38970.000000   18287.00000


E. we used head(20) to display the first  20 records in the dataset.

In [15]:
print(retail_frame.head(20))

   InvoiceNo StockCode  ... CustomerID         Country
0     536365    85123A  ...    17850.0  United Kingdom
1     536365     71053  ...    17850.0  United Kingdom
2     536365    84406B  ...    17850.0  United Kingdom
3     536365    84029G  ...    17850.0  United Kingdom
4     536365    84029E  ...    17850.0  United Kingdom
5     536365     22752  ...    17850.0  United Kingdom
6     536365     21730  ...    17850.0  United Kingdom
7     536366     22633  ...    17850.0  United Kingdom
8     536366     22632  ...    17850.0  United Kingdom
9     536367     84879  ...    13047.0  United Kingdom
10    536367     22745  ...    13047.0  United Kingdom
11    536367     22748  ...    13047.0  United Kingdom
12    536367     22749  ...    13047.0  United Kingdom
13    536367     22310  ...    13047.0  United Kingdom
14    536367     84969  ...    13047.0  United Kingdom
15    536367     22623  ...    13047.0  United Kingdom
16    536367     22622  ...    13047.0  United Kingdom
17    5363

F. we  used sample to extract the random rows in the dataset then assigned .30 in the fraq to specify that we are only getting .30 or 30%

In [18]:
 retail_sample_frame = retail_frame.sample(frac = .30)

G.we used .unique() to show only the unique values

In [22]:
print(retail_frame['Description'].unique())

['WHITE HANGING HEART T-LIGHT HOLDER' 'WHITE METAL LANTERN'
 'CREAM CUPID HEARTS COAT HANGER' ... 'lost'
 'CREAM HANGING HEART T-LIGHT HOLDER' 'PAPER CRAFT , LITTLE BIRDIE']


H. we used groupby to indicate that we are grouping the dataframe in to dates and customer id then inside the groupby we used grouper to indicate to sepecify what frequency of the date we want to show in this case we want to show the weeks. then we indicated that we are calculating the sum of the unit price per group

In [34]:
print(retail_frame.groupby([pd.Grouper(key='InvoiceDate',freq='W'),'CustomerID']).UnitPrice.sum().head(50))

InvoiceDate  CustomerID
2010-12-05   12395.0        37.41
             12427.0        49.25
             12431.0        73.90
             12433.0       102.67
             12471.0         8.50
             12472.0       217.46
             12557.0         7.85
             12567.0       287.19
             12583.0        55.29
             12600.0        19.95
             12647.0       125.15
             12662.0        44.37
             12682.0        67.07
             12686.0        22.81
             12712.0        46.64
             12720.0        80.42
             12725.0        98.50
             12738.0        43.45
             12747.0        21.13
             12748.0       452.04
             12763.0        28.99
             12779.0        18.99
             12791.0        16.85
             12793.0        38.40
             12838.0       116.43
             12841.0       129.29
             12855.0         4.65
             12868.0        38.05
             12915.0    

I. we used groupby to indicate that we are grouping the dataframe in to dates and description then inside the groupby we used grouper to indicate to sepecify what frequency of the date we want to show in this case we want to show the weeks. then we indicated that we are calculating the sum of the unit price per group

In [33]:
print(retail_frame.groupby([pd.Grouper(key='InvoiceDate',freq='W'),'Description']).UnitPrice.sum().head(50))

InvoiceDate  Description                        
2010-12-05    4 PURPLE FLOCK DINNER CANDLES           5.10
              OVAL WALL MIRROR DIAMANTE              20.38
              SET 2 TEA TOWELS I LOVE LONDON         67.88
             10 COLOUR SPACEBOY PEN                  14.33
             12 COLOURED PARTY BALLOONS               1.95
             12 DAISY PEGS IN WOOD BOX                8.25
             12 IVORY ROSE PEG PLACE SETTINGS         6.26
             12 MESSAGE CARDS WITH ENVELOPES         19.98
             12 PENCIL SMALL TUBE WOODLAND            8.39
             12 PENCILS SMALL TUBE RED RETROSPOT     12.92
             12 PENCILS SMALL TUBE SKULL             11.01
             12 PENCILS TALL TUBE POSY                1.70
             12 PENCILS TALL TUBE RED RETROSPOT       5.10
             12 PENCILS TALL TUBE SKULLS              5.95
             12 PENCILS TALL TUBE WOODLAND            6.72
             12 PINK ROSE PEG PLACE SETTINGS          1.25
       

J. we used group by and specified that we are adding the unitprice per group then we stored it to the new object called group_by_country

In [32]:
group_by_country = retail_frame.groupby('Country').UnitPrice.sum()
print(group_by_country)

Country
Australia               4.054750e+03
Austria                 1.701520e+03
Bahrain                 8.657000e+01
Belgium                 7.540130e+03
Brazil                  1.426000e+02
Canada                  9.105800e+02
Channel Islands         3.738550e+03
Cyprus                  3.920070e+03
Czech Republic          8.815000e+01
Denmark                 1.266950e+03
EIRE                    4.844719e+04
European Community      2.940500e+02
Finland                 3.786850e+03
France                  4.303199e+04
Germany                 3.766600e+04
Greece                  7.132900e+02
Hong Kong               1.224150e+04
Iceland                 4.812100e+02
Israel                  1.079040e+03
Italy                   3.879390e+03
Japan                   8.148600e+02
Lebanon                 2.424400e+02
Lithuania               9.944000e+01
Malta                   6.660100e+02
Netherlands             6.492550e+03
Norway                  6.529060e+03
Poland                  1.4222

K.we used group by and specified that we are adding the unitprice per group then we stored it to the new object called group_by_description

In [27]:
group_by_description = retail_frame.groupby('Description').UnitPrice.sum()
print(group_by_description)

Description
20713                             -400
 4 PURPLE FLOCK DINNER CANDLES     144
 50'S CHRISTMAS GIFT BAG LARGE    1913
 DOLLY GIRL BEAKER                2448
 I LOVE LONDON MINI BACKPACK       389
                                  ... 
wrongly marked carton 22804       -256
wrongly marked. 23343 in box     -3100
wrongly sold (22719) barcode       170
wrongly sold as sets              -600
wrongly sold sets                 -975
Name: Quantity, Length: 4223, dtype: int64


**Part 1 Summary**

in the part1, we preprocessed the excel file called, Online Retail.xlsx.
in the preprocessing we cleaned the excel file by finding all the missing values then replaced it with its median and mode, we also viewd its statistical details and selected 30 percent of its row size as samples. we also identified all of the unique description feature and grouped them to view the ammount added pero week then we created objects that contains all the columns together with its total unit price.


# Part 2

A.using panda's read_excel we imported Family Income and Expenditure.xlsx

In [36]:
family_frame = pd.read_excel('Family Income and Expenditure.xlsx')

B.using isnull() we identify the missing values and we used sum to get the total missing values

In [43]:
print(family_frame.isnull().sum())

Total Household Income                              0
Region                                              0
Total Food Expenditure                              0
Main Source of Income                               0
Agricultural Household indicator                    0
Bread and Cereals Expenditure                      15
Total Rice Expenditure                              0
Meat Expenditure                                    0
Total Fish and  marine products Expenditure         0
Fruit Expenditure                                   0
Vegetables Expenditure                              0
Restaurant and hotels Expenditure                  19
Alcoholic Beverages Expenditure                     0
Tobacco Expenditure                                 0
Clothing, Footwear and Other Wear Expenditure       0
Housing and water Expenditure                       0
Imputed House Rental Value                          0
Medical Care Expenditure                            0
Transportation Expenditure  

C.we segregated the 3 numerical and 2 categorical features then replaced the missing values using fillna() then indicated that we are replacing it with the mean and the mode of each features.

In [44]:
num1 = family_frame['Bread and Cereals Expenditure']
num2 = family_frame['Restaurant and hotels Expenditure']
num3 = family_frame['Household Head Age']
obj1 = family_frame['Household Head Occupation']
obj2 = family_frame['Household Head Class of Worker']

num1.fillna(num1.mean(),inplace=True)
num2.fillna(num2.mean(),inplace=True)
num3.fillna(num3.mean(),inplace=True)
obj1.fillna(obj1.mode(),inplace=True)
obj2.fillna(obj2.mode(),inplace=True)


family_frame['Bread and Cereals Expenditure'] = num1
family_frame['Restaurant and hotels Expenditure'] = num2 
family_frame['Household Head Age'] = num3
family_frame['Household Head Occupation'] = obj1  
family_frame['Household Head Class of Worker'] = obj2

D.we printed the firs 20 values of the frame using head()

In [47]:
print(family_frame.head(20))

    Total Household Income  ... Number of Motorcycle/Tricycle
0                   480332  ...                             1
1                   198235  ...                             2
2                    82785  ...                             0
3                   107589  ...                             0
4                   189322  ...                             1
5                   152883  ...                             1
6                   198621  ...                             1
7                   134961  ...                             1
8                   171152  ...                             0
9                   625753  ...                             0
10                  147005  ...                             1
11                  253130  ...                             0
12                  101111  ...                             0
13                  128214  ...                             1
14                   91525  ...                             0
15      

E. we  used sample to extract the random rows in the dataset then assigned .30 in the fraq to specify that we are only getting .20 or 20%

In [48]:
 family_sample_frame = family_frame.sample(frac = .20)

F. we created bins to specify the intervals to be used, then used pd,cut to segregate the data into their respective bins

In [51]:
bins = [0,25,50,75,100]

data = family_frame['Household Head Age']

cat = pd.cut(data,bins)




G.we used pd.value_counts to print the occurence per bin

In [53]:
print(pd.value_counts(cat), '\n')

(25, 50]     19798
(50, 75]     18635
(75, 100]     2274
(0, 25]        837
Name: Household Head Age, dtype: int64 



H. we used q cut to automatcally set the bin intervals and balance out the number of occurance and specified the quantile to 5 so that we can have only 5 bins

In [55]:
qcat = pd.qcut(data,q=5)


I. we used pd.value_counts to print the occurence per bin

In [56]:
print(pd.value_counts(qcat), '\n')

(8.999, 39.0]    8997
(47.0, 55.0]     8682
(39.0, 47.0]     8295
(55.0, 64.0]     7819
(64.0, 99.0]     7751
Name: Household Head Age, dtype: int64 



J. we used copy() to copy the dataframe in to the assigned new object

In [57]:
family_sample_frame_copy = family_sample_frame.copy()

K. we used select dtypes to specify that we are only needing the columns that has  int64 and fload 64.

In [68]:
dt = family_sample_frame_copy.select_dtypes(include=['int64'])
dt2 = family_sample_frame_copy.select_dtypes(include=['float64'])

types = dt + dt2
print(types.columns.values)


['Agricultural Household indicator' 'Alcoholic Beverages Expenditure'
 'Bread and Cereals Expenditure'
 'Clothing, Footwear and Other Wear Expenditure'
 'Communication Expenditure' 'Crop Farming and Gardening expenses'
 'Education Expenditure' 'Electricity' 'Fruit Expenditure' 'House Age'
 'House Floor Area' 'Household Head Age' 'Housing and water Expenditure'
 'Imputed House Rental Value' 'Meat Expenditure'
 'Medical Care Expenditure' 'Members with age 5 - 17 years old'
 'Members with age less than 5 year old'
 'Miscellaneous Goods and Services Expenditure' 'Number of Airconditioner'
 'Number of CD/VCD/DVD' 'Number of Car, Jeep, Van'
 'Number of Cellular phone' 'Number of Component/Stereo set'
 'Number of Landline/wireless telephones' 'Number of Motorcycle/Tricycle'
 'Number of Motorized Banca' 'Number of Personal Computer'
 'Number of Refrigerator/Freezer' 'Number of Stove with Oven/Gas Range'
 'Number of Television' 'Number of Washing Machine' 'Number of bedrooms'
 'Restaurant and h

after getting the names of the columns that has float and int we used loc to locate then and to get their values

In [70]:
x = family_sample_frame_copy.loc[:,types.columns].values



L Standardize,  We imported StandardScaler from sklearn to use the StandardScaler function to standardize our data , to do this we assigned scaler as the standarscaler() then used it to call fit_transform and feed the data which is x to get all the standardized value, after that we generate feature names from 0 to n then we combined it all to form a datafame

In [73]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

x = scaler.fit_transform(x) 

x_features = ['feature'+str(i) for i in range(x.shape[1])]

normalized_df = pd.DataFrame(x,columns=x_features)
print(normalized_df)

      feature0  feature1   feature2  ...  feature42  feature43  feature44
0     0.830116  1.536673  -0.219781  ...  -0.229003   0.066989   0.248761
1    -0.639299 -0.498034  -0.592754  ...  -1.086888  -0.281829  -1.042784
2    -0.639299 -0.283384  -0.743997  ...   1.486766   0.072781   0.116730
3    -0.639299  0.843531   2.340850  ...   4.060419   1.436196   1.579370
4     0.830116 -0.498034  -0.141401  ...  -1.086888  -0.578518  -0.780841
...        ...       ...        ...  ...        ...        ...        ...
8304 -0.639299 -0.225249  -0.726222  ...  -1.086888  -0.490348  -0.416846
8305 -0.639299  2.739609   1.190823  ...   2.344650  -0.131876  -1.027642
8306  2.299531  0.548387  -0.384278  ...  -0.229003  -0.233561  -0.101909
8307 -0.639299 -0.377293  17.489032  ...   0.628881   0.091445  -0.487101
8308 -0.639299  0.615465   0.324280  ...  -0.229003  -0.135416   0.690884

[8309 rows x 45 columns]


M. PCA, we imported PCA from sklearn then created pca_data to call the pca function that indicates we only need 2 components then we stored the pca result to pca_res1 then created the dataframe using pca_res2

In [75]:
from sklearn.decomposition import PCA
pca_data = PCA(n_components=2)
pca_res1 = pca_data.fit_transform(x)
pca_res2 = pd.DataFrame(data = pca_res1, columns=['PC1','PC2'])
print(pca_res2)

            PC1       PC2
0      1.088612 -0.130079
1      3.469055 -3.722648
2      4.392958  0.305440
3      3.587831  4.660023
4     -2.187468 -0.331557
...         ...       ...
8304  -1.109266 -2.172807
8305  -0.625801  3.779036
8306  -0.236532 -0.558353
8307  10.331835  4.328220
8308   1.153895  0.399577

[8309 rows x 2 columns]


N. using explain_variance_ratio_ to our pca_data, we can see the explaied variance ratio for each component

In [76]:
print('Explained variation per principal component: {}'.format(pca_data.explained_variance_ratio_))

Explained variation per principal component: [0.26051009 0.08108816]


O. we copied the columns A, B, C, D, Y, Z, AA, and AB by using iloc[] to specify the number of index of the column then stored it in a copy object

In [118]:
family_sample_frame_copy2 = family_sample_frame.iloc[:,[0,1,2,3,24,26,27]]


print(family_sample_frame_copy2)



       Total Household Income  ... Household Head Marital Status
29842                  204078  ...                       Married
16752                  510235  ...            Divorced/Separated
36403                  482727  ...                       Married
35004                  346033  ...                       Married
19914                   88005  ...                       Widowed
...                       ...  ...                           ...
4799                   166529  ...                       Married
9285                   260715  ...                       Married
37526                  157616  ...                       Married
10363                  727146  ...                       Widowed
13223                  163266  ...                       Married

[8309 rows x 7 columns]


P. we imported LabelEncoder from sk learn to feature encode the region column then stored it in the new column called B

In [119]:
from sklearn.preprocessing import LabelEncoder

encoder = LabelEncoder()

family_sample_frame_copy2['B'] = encoder.fit_transform(family_sample_frame['Region'])


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """


Q. We imported LabelBinarizer from sklearn then stored it in the binarizer 

we assigned x3 to store our  selected columns and its values using iloc then we sepecified the index of the columns. we then reshaped the x3 to a 1D array then using binarizer.fit_transform we feature encode the columns specifed then the result of it is stored in the binarizer_res

In [200]:
from sklearn.preprocessing import LabelBinarizer

binarizer = LabelBinarizer()

x3 = family_sample_frame.iloc[:,[3,25,27]].values



x3 = x3.reshape(-1,1)

print(x3)
binarizer_res = binarizer.fit_transform(x3)

print(binarizer_res)



[['Enterpreneurial Activities']
 ['Male']
 ['Married']
 ...
 ['Enterpreneurial Activities']
 ['Male']
 ['Married']]
[[0 0 1 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 1 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]


R. to concatinate all the data we first reshaped the binarizer_res to its original shape which has 3 columns and 83090 rows, then we created a dataframe for the binarizer_res that has 3 column headers b1 to b3, then we combined the family_sample_frame_copy2 with the new frame newd in an object, then using final we concatinate the 2 dataframes and specified the axis to prevent the added frame values to be none and we joined it using innerjoin

In [199]:
binarizer_res = binarizer_res.reshape(83090,3)

print(binarizer_res)

newd= pd.DataFrame(data=binarizer_res, columns= ['b1','b2','b3'])


combine = [family_sample_frame_copy2,newd]
final = pd.concat(combine,axis=1,join='inner')
print(final.head(50))

[[0 0 1]
 [0 0 0]
 [0 0 0]
 ...
 [0 0 0]
 [0 1 0]
 [0 0 0]]
       Total Household Income                     Region  ...  b2 b3
29842                  204078        II - Cagayan Valley  ...   0  0
16752                  510235           IVA - CALABARZON  ...   0  0
36403                  482727  IX - Zasmboanga Peninsula  ...   0  0
35004                  346033        III - Central Luzon  ...   0  1
19914                   88005                       ARMM  ...   1  0
29466                  129845           IVA - CALABARZON  ...   0  0
28775                  209413          I - Ilocos Region  ...   0  0
18371                  109078                       ARMM  ...   0  0
10597                   23151      VII - Central Visayas  ...   0  0
339                    201271                     Caraga  ...   0  1
35652                   94613  IX - Zasmboanga Peninsula  ...   0  0
14595                  174520       VI - Western Visayas  ...   0  0
8428                   316594           IVA

**Part2**

In the part2 we preprocessed Family Income and Expenditure.xlsx. We filled the missing values with the mean and mode of their respective columns and we get random samples from the 20% of the population. We also segregated the household head age and created 4 bins with the range of 0,25,50,75,100 then assigned them to their respective bins and using q cut we automatically generated the intervals and balanced the number of occurances. in this part we replicated our dataframe in preperation for the principal component analysis or pca and extracted only 2 principal components as a result our Explained variation per principal component are [0.26051009 0.08108816]. we again prepared that data by copying it in preperation for feature encoding using labelencoder and labelbinarier from sklearn libraries, in the process we feature encoded the regions column then we used labelbinarizer to featrure encode the columns, Main Source of Income, Household Head Sex and Household Head Marital Status and finally concatinated their results and displayed only 50 of records.
